File: packages/sdk-react-native/ios/SalvePushStorage.swift
// Filesystem layout for installed releases (ADR 0004): releases/, state.json, mount.marker, crash.marker.
import Foundation

enum SalvePushStorage {
  static func baseDirectory() -> URL {
    let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
    let dir = appSupport.appendingPathComponent("salve-push", isDirectory: true)
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    return dir
  }

  static func releasesDirectory() -> URL {
    baseDirectory().appendingPathComponent("releases", isDirectory: true)
  }

  static func bundlePath(for releaseId: String) -> URL {
    releasesDirectory().appendingPathComponent(releaseId, isDirectory: true).appendingPathComponent("bundle.js")
  }

  static func statePath() -> URL {
    baseDirectory().appendingPathComponent("state.json")
  }

  static func mountMarkerPath() -> URL {
    baseDirectory().appendingPathComponent("mount.marker")
  }

  static func crashMarkerPath() -> URL {
    baseDirectory().appendingPathComponent("crash.marker")
  }

  static func readState() -> SalvePushState {
    guard let data = try? Data(contentsOf: statePath()),
      let state = try? JSONDecoder().decode(SalvePushState.self, from: data)
    else {
      return SalvePushState(currentReleaseId: nil, previousReleaseId: nil, bootStatus: "confirmed")
    }
    return state
  }

  static func writeState(_ state: SalvePushState) throws {
    let data = try JSONEncoder().encode(state)
    try data.write(to: statePath(), options: .atomic)
  }

  static func markMounted() {
    FileManager.default.createFile(atPath: mountMarkerPath().path, contents: nil)
  }

  static func clearMountedMarker() {
    try? FileManager.default.removeItem(at: mountMarkerPath())
  }

  static func isMounted() -> Bool {
    FileManager.default.fileExists(atPath: mountMarkerPath().path)
  }

  static func installUpdate(releaseId: String, bundle: Data) throws {
    let releaseDir = releasesDirectory().appendingPathComponent(releaseId, isDirectory: true)
    try FileManager.default.createDirectory(at: releaseDir, withIntermediateDirectories: true)
    let finalPath = releaseDir.appendingPathComponent("bundle.js")
    let tempPath = releaseDir.appendingPathComponent("bundle.js.tmp")
    try bundle.write(to: tempPath, options: .atomic)
    if FileManager.default.fileExists(atPath: finalPath.path) {
      try FileManager.default.removeItem(at: finalPath)
    }
    try FileManager.default.moveItem(at: tempPath, to: finalPath)

    var state = readState()
    let previous = state.currentReleaseId
    state.previousReleaseId = previous
    state.currentReleaseId = releaseId
    state.bootStatus = "pending"
    try writeState(state)
    clearMountedMarker()

    pruneOldReleases(keep: [releaseId, previous].compactMap { $0 })
  }

  static func confirmBoot() throws {
    markMounted()
    var state = readState()
    state.bootStatus = "confirmed"
    try writeState(state)
  }

  static func getCurrentReleaseId() -> String {
    readState().currentReleaseId ?? ""
  }

  /// Called by bundleURL()/getJSBundleFile() before JS boots. Returns the
  /// bundle path to use, or nil to fall back to the app's embedded bundle.
  static func resolveBundlePath() -> String? {
    processCrashMarkerIfPresent()
    guard let currentId = readState().currentReleaseId else { return nil }
    let path = bundlePath(for: currentId)
    guard FileManager.default.fileExists(atPath: path.path) else { return nil }
    return path.path
  }

  private static func processCrashMarkerIfPresent() {
    guard let data = try? Data(contentsOf: crashMarkerPath()) else { return }
    try? FileManager.default.removeItem(at: crashMarkerPath())
    guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      json["isAutoRollback"] as? Bool == true
    else { return }
    rollbackToPrevious()
  }

  /// Called by the same-session exception handler when a crash happens
  /// before the current release confirmed boot.
  static func rollbackToPreviousIfUnconfirmed() {
    guard !isMounted() else { return }
    rollbackToPrevious()
  }

  private static func rollbackToPrevious() {
    var state = readState()
    guard let previous = state.previousReleaseId else { return }
    state.currentReleaseId = previous
    state.previousReleaseId = nil
    state.bootStatus = "confirmed"
    try? writeState(state)
    markMounted()
  }

  private static func pruneOldReleases(keep: [String]) {
    guard
      let entries = try? FileManager.default.contentsOfDirectory(
        at: releasesDirectory(), includingPropertiesForKeys: nil)
    else { return }
    for entry in entries where !keep.contains(entry.lastPathComponent) {
      try? FileManager.default.removeItem(at: entry)
    }
  }
}

File: packages/sdk-react-native/android/src/main/java/com/margelo/nitro/salvepush/SalvePushStorage.kt
// Filesystem layout for installed releases (ADR 0004): releases/, state.json, mount.marker, crash.marker.
// Every function takes Context explicitly: getJSBundleFile() runs before the
// React bridge (and therefore NitroModules.applicationContext) exists.
package com.margelo.nitro.salvepush

import android.content.Context
import org.json.JSONObject
import java.io.File

object SalvePushStorage {
  fun baseDirectory(context: Context): File {
    val dir = File(context.filesDir, "salve-push")
    dir.mkdirs()
    return dir
  }

  private fun releasesDirectory(context: Context): File {
    val dir = File(baseDirectory(context), "releases")
    dir.mkdirs()
    return dir
  }

  private fun bundlePath(context: Context, releaseId: String): File =
    File(File(releasesDirectory(context), releaseId), "bundle.js")

  private fun statePath(context: Context): File = File(baseDirectory(context), "state.json")

  private fun mountMarkerPath(context: Context): File = File(baseDirectory(context), "mount.marker")

  private fun crashMarkerPath(context: Context): File = File(baseDirectory(context), "crash.marker")

  private fun readState(context: Context): SalvePushState {
    return try {
      SalvePushState.fromJson(statePath(context).readText())
    } catch (e: Exception) {
      SalvePushState.confirmedEmpty
    }
  }

  private fun writeState(context: Context, state: SalvePushState) {
    val tmp = File(baseDirectory(context), "state.json.tmp")
    tmp.writeText(state.toJson())
    tmp.renameTo(statePath(context))
  }

  private fun markMounted(context: Context) {
    mountMarkerPath(context).createNewFile()
  }

  private fun clearMountedMarker(context: Context) {
    mountMarkerPath(context).delete()
  }

  fun isMounted(context: Context): Boolean = mountMarkerPath(context).exists()

  fun installUpdate(context: Context, releaseId: String, bundle: ByteArray) {
    val releaseDir = File(releasesDirectory(context), releaseId)
    releaseDir.mkdirs()
    val finalFile = File(releaseDir, "bundle.js")
    val tempFile = File(releaseDir, "bundle.js.tmp")
    tempFile.writeBytes(bundle)
    if (finalFile.exists()) finalFile.delete()
    tempFile.renameTo(finalFile)

    val previousState = readState(context)
    val previous = previousState.currentReleaseId
    writeState(
      context,
      previousState.copy(
        previousReleaseId = previous,
        currentReleaseId = releaseId,
        bootStatus = "pending",
      ),
    )
    clearMountedMarker(context)

    pruneOldReleases(context, listOfNotNull(releaseId, previous))
  }

  fun confirmBoot(context: Context) {
    markMounted(context)
    writeState(context, readState(context).copy(bootStatus = "confirmed"))
  }

  fun getCurrentReleaseId(context: Context): String =
    readState(context).currentReleaseId ?: ""

  /** Called by getJSBundleFile() before JS boots. Returns null to fall back to the embedded bundle. */
  fun resolveBundlePath(context: Context): String? {
    processCrashMarkerIfPresent(context)
    val currentId = readState(context).currentReleaseId ?: return null
    val path = bundlePath(context, currentId)
    return if (path.exists()) path.absolutePath else null
  }

  private fun processCrashMarkerIfPresent(context: Context) {
    val marker = crashMarkerPath(context)
    if (!marker.exists()) return
    val content = try {
      JSONObject(marker.readText())
    } catch (e: Exception) {
      null
    }
    marker.delete()
    if (content?.optBoolean("isAutoRollback", false) == true) {
      rollbackToPrevious(context)
    }
  }

  /** Called by the same-session exception handler when a crash happens before the current release confirmed boot. */
  fun rollbackToPreviousIfUnconfirmed(context: Context) {
    if (!isMounted(context)) {
      rollbackToPrevious(context)
    }
  }

  private fun rollbackToPrevious(context: Context) {
    val state = readState(context)
    val previous = state.previousReleaseId ?: return
    writeState(
      context,
      state.copy(currentReleaseId = previous, previousReleaseId = null, bootStatus = "confirmed"),
    )
    markMounted(context)
  }

  private fun pruneOldReleases(context: Context, keep: List<String>) {
    releasesDirectory(context).listFiles()?.forEach { entry ->
      if (!keep.contains(entry.name)) {
        entry.deleteRecursively()
      }
    }
  }
}
